src/main/java/gregtech/common/metatileentities/storage/MetaTileEntityQuantumStorage.java

```diff
@@ -0,0 +1,190 @@
+
+    public MetaTileEntityQuantumStorage(ResourceLocation metaTileEntityId) {
+        super(metaTileEntityId);
+    }
+
+    @Override
+    public void setConnected(IQuantumController controller) {
+        if (getWorld().isRemote) return;
+        if (!controller.getPos().equals(controllerPos)) {
+            this.controller = new WeakReference<>(controller);
+            this.controllerPos = controller.getPos();
+            if (!getWorld().isRemote) {
+                writeCustomData(UPDATE_CONTROLLER_POS, buf -> buf.writeBlockPos(controllerPos));
+                markDirty();
+            }
+        }
+    }
+
+    @Override
+    public void setDisconnected() {
+        if (getWorld().isRemote) return;
+        if (controllerPos != null && !tryFindNewNetwork()) {
+            // first try to find a new adjacent network
+            controller = null;
+            controllerPos = null;
+
+            writeCustomData(REMOVE_CONTROLLER, buf -> {});
+            markDirty();
+        }
+    }
+
+    // use this to make sure controller is properly initialized
+    @Override
+    public IQuantumController getController() {
+        if (getWorld().isRemote) return null; // quick safety check
+        if (isConnected()) {
+            if (controller != null && controller.get() != null) return controller.get();
+            MetaTileEntity mte = GTUtility.getMetaTileEntity(getWorld(), controllerPos);
+            if (mte instanceof IQuantumController) {
+                controller = new WeakReference<>((IQuantumController) mte);
+                return (IQuantumController) mte;
+            } else {
+                // controller is no longer there for some reason, need to disconnect
+                controller = null;
+                controllerPos = null;
+                writeCustomData(REMOVE_CONTROLLER, buf -> {});
+                markDirty();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isConnected() {
+        // use controllerPos here because it is synced
+        // on both sides, where controller is not
+        return controllerPos != null;
+    }
+
+    @Override
+    public BlockPos getControllerPos() {
+        return controllerPos;
+    }
+
+    @Override
+    public void onRemoval() {
+        if (!getWorld().isRemote && isConnected()) {
+            IQuantumController controller = getController();
+            if (controller != null) controller.rebuildNetwork();
+        }
+    }
+
+    // todo might not be good time to do this
+    // todo need to make sure it is AFTER nbt has been read
+    @Override
+    public void onPlacement() {
+        // add to the network if an adjacent block is part of a network
+        // use whatever we find first, merging networks is not supported
+        if (!getWorld().isRemote) {
+            tryFindNewNetwork();
+        }
+    }
+
+    private boolean tryFindNewNetwork() {
+        for (EnumFacing facing : EnumFacing.VALUES) {
```

Comment:
This could probably also be checked in `onNeighborChanged`, so that the case that I presented in discord could be fixed a bit better.

Or an option to use the Neighbor cache here.